paasta_tools/contrib/check_orphans.py

```diff
@@ -0,0 +1,213 @@
+        if service in blacklisted_services:
+            continue
+        service_instances = zk.get_children(os.path.join(PREFIX, service))
+        instances_data = {}
+        for instance in service_instances:
+            instance_node = zk.get(os.path.join(PREFIX, service, instance))
+            instances_data[instance] = json.loads(instance_node[0])
+            zk_data[service] = instances_data
+
+    return zk_data
+
+
+class InstanceTuple(NamedTuple):
+    host: str
+    port: int
+    service: str
+
+
+def read_from_zk_data(registrations: SmartstackData) -> Set[InstanceTuple]:
+    return {
+        InstanceTuple(instance_data['host'], instance_data['port'], service)
+        for service, instance in registrations.items()
+        for instance_data in instance.values()
+    }
+
+
+async def transfer_one_file(host: str, port: int = DEFAULT_NERVE_XINETD_PORT) -> Tuple[str, Optional[str]]:
+    logging.debug(f'getting file from {host}')
+    try:
+        reader, _ = await asyncio.wait_for(
+            asyncio.open_connection(host=host, port=port, limit=2**32),
+            timeout=1.0,
+        )
+        resp = await asyncio.wait_for(reader.read(), timeout=1.0)
+    except (asyncio.TimeoutError, ConnectionRefusedError):
+        logging.error(f'error getting file from {host}')
+        return (host, None)
+
+    return (host, resp.decode())
+
+
+async def gather_files(hosts: Set[str]) -> Dict[str, str]:
+    logging.info('gathering files from {} hosts'.format(len(hosts)))
+    tasks = [transfer_one_file(host) for host in hosts]
+    responses = {}
+    for idx in range(0, len(tasks), CHUNK_SIZE):
+        resp = await asyncio.gather(*tasks[idx:idx + CHUNK_SIZE], return_exceptions=True)
+        responses.update(dict(resp))
+    return responses
+
+
+def read_one_nerve_file(nerve_config: str) -> Set[InstanceTuple]:
+    services = json.loads(nerve_config)['services']
+    return {
+        InstanceTuple(service['host'], service['port'], service['zk_path'][len(PREFIX):])
+        for service in services.values()
+        if service['zk_path'].startswith(PREFIX)
+    }
+
+
+def read_nerve_files(nerve_configs: Dict[str, Optional[str]]) -> Tuple[Set[InstanceTuple], Set[str]]:
+    instance_set: Set[InstanceTuple] = set()
+    not_found_hosts: Set[str] = set()
+    for host, host_config in nerve_configs.items():
+        if host_config is None:
+            not_found_hosts.add(host)
+        else:
+            instance_set |= read_one_nerve_file(host_config)
+    return instance_set, not_found_hosts
+
+
+def get_instance_data(service_blacklist_path: str) -> Tuple[Set[InstanceTuple], Set[InstanceTuple]]:
+    blacklisted_services = set(yaml.load(service_blacklist_path))
+    # Dump ZK
+    zk_data = get_zk_data(blacklisted_services)
+    zk_instance_data = read_from_zk_data(zk_data)
+
+    hosts = {x[0] for x in zk_instance_data}
+
+    # Dump Nerve configs from each host via xinetd
+    coro = gather_files(hosts)
```

Comment:
What is coro?

Reply:
`coroutine`. it can be combined with the next line, which i'll do now